realm/realm-library/src/main/java/io/realm/internal/OsSharedRealm.java

```diff
@@ -160,6 +160,14 @@ protected PartialSyncCallback(String className) {
     // JNI will only hold a weak global ref to this.
     public final RealmNotifier realmNotifier;
     public final Capabilities capabilities;
+    // For the Java callbacks during constructing in Object Store, some temporary OsSharedRealm objects need to be
+    // created as the parameter of the callback. The native pointers of those temp OsSharedRealm objects have to be
+    // valid during the whole life cycle of the Java object. The living native pointers still hold a ref-count to the
+    // SharedRealm which means the SharedRealm won't be closed automatically if there is any exception throws during
+    // construction. GC will clear them later, but that would be too late. So we are tracking the temp OsSharedRealm
+    // during the construction stage and manually close them if exception throws.
+    private final static List<OsSharedRealm> sharedRealmsUnderConstruction = new CopyOnWriteArrayList<OsSharedRealm>();
```

Comment:
Does this need to be a `CopyOnWriteArrayList` ? If I remember correctly are all `Realm.getInstance()` methods under a synchronized lock before we hit this part of the code?

Reply:
nope, we only lock on the same configuration now . see https://github.com/realm/realm-java/pull/4551